module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,640 @@
+)
+
+// ErrRequesterHalted is returned when an invalid ExectutionData is encountered
+var ErrRequesterHalted = errors.New("requester was halted due to invalid data")
+
+// ExecutionDataConfig contains configuration options for the ExecutionDataRequester
+type ExecutionDataConfig struct {
+	// The first block height for which to request ExecutionData
+	StartBlockHeight uint64
+
+	// Max number of ExecutionData objects to keep when waiting to send notifications.
+	// Dropped data is refetched from disk.
+	MaxCachedEntries uint64
+
+	// Max number of unsent notifications to allow before pausing new fetches. After exceeding this
+	// limit, the requester will stop processing new finalized block notifications. This prevents
+	// unbounded memory use by the requester if it gets stuck fetching a specific height.
+	MaxSearchAhead uint64
+
+	// The timeout for fetching ExecutionData from the db/network
+	FetchTimeout time.Duration
+
+	// Exponential backoff settings for download retries
+	RetryDelay    time.Duration
+	MaxRetryDelay time.Duration
+
+	// Whether or not to run datastore check on startup
+	CheckEnabled bool
+}
+
+type executionDataRequester struct {
+	component.Component
+	cm      *component.ComponentManager
+	ds      datastore.Batching
+	bs      network.BlobService
+	eds     state_synchronization.ExecutionDataService
+	metrics module.ExecutionDataRequesterMetrics
+	log     zerolog.Logger
+
+	// Local db objects
+	blocks  storage.Blocks
+	results storage.ExecutionResults
+
+	status *status.Status
+	config ExecutionDataConfig
+
+	// Notifiers for queue consumers
+	blockNotifier engine.Notifier
+
+	// Job queues
+	blockConsumer        *jobqueue.WrappedConsumer
+	notificationConsumer *jobqueue.WrappedConsumer
+
+	// List of callbacks to call when ExecutionData is successfully fetched for a block
+	consumers []state_synchronization.ExecutionDataReceivedCallback
+
+	consumerMu sync.RWMutex
+}
+
+var _ state_synchronization.ExecutionDataRequester = (*executionDataRequester)(nil)
+
+// New creates a new execution data requester component
+func New(
+	log zerolog.Logger,
+	edrMetrics module.ExecutionDataRequesterMetrics,
+	datastore datastore.Batching,
+	blobservice network.BlobService,
+	eds state_synchronization.ExecutionDataService,
+	processedHeight storage.ConsumerProgress,
+	processedNotifications storage.ConsumerProgress,
+	state protocol.State,
+	blocks storage.Blocks,
+	results storage.ExecutionResults,
+	cfg ExecutionDataConfig,
+) (state_synchronization.ExecutionDataRequester, error) {
+	e := &executionDataRequester{
+		log:           log.With().Str("component", "execution_data_requester").Logger(),
+		ds:            datastore,
+		bs:            blobservice,
+		eds:           eds,
+		metrics:       edrMetrics,
+		blocks:        blocks,
+		results:       results,
+		config:        cfg,
+		blockNotifier: engine.NewNotifier(),
+	}
+
+	executionDataNotifier := engine.NewNotifier()
+	rootHeight := e.config.StartBlockHeight - 1
```

Comment:
for the requester, I want to allow the node operator to specify the start height so they don't always have to sync all blocks for the whole spork. This is a simplified way to deal with expiring data before we build that system

Reply:
I'm just afraid that there could possibly create "gaps" in the synced data.

For instance, I start with `StartBlockHeight: 100`, and when synced to height 200, I restart and change the flag to be `StartBlockHeight: `1000`, and then synced to height `1200`, then I will have a gap between height `201` to `999`, where I have no execution data.

I'm not sure if some code would assume there is no gap, if so, these gaps would break the data integrity.

Reply:
Since the job queue state is stored in the same db as the execution data, I think it should always catch up from where it left off. The only case where the start block takes effect is when starting from an empty execution data db. 

The start height is useful for operators if they want to prune their db midspork. They could pick some block height they've already synced, stop their node, delete the execution data db, then start from the new height. Eventually, we'll provide features to automatically prune data

Reply:
I'm ok changing the naming, and would prefer to add the check. I definitely agree from the code perspective the -1 isn't great, but I'm thinking about this from the perspective of the operator configuring it.

They would either use the default which is the root block height from the `root-protocol-state-snapshot.json` file, or they'd specify the first block they want to sync. I think it would be confusing for an operator to have to pick the first block they want and subtract 1, and even harder to describe succinctly in a CLI tool tip.